pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/ReplicaGroupInstanceSelector.java

```diff
@@ -152,21 +154,28 @@ private Pair<Map<String, String>, Map<String, String>> selectServersUsingAdaptiv
       SegmentInstanceCandidate selectedInstance = candidates.get(instanceIdx);
       // Adaptive Server Selection
       // TODO: Support numReplicaGroupsToQuery with Adaptive Server Selection.
-      if (!serverRankList.isEmpty()) {
-        int minIdx = Integer.MAX_VALUE;
+      if (!serverRankMap.isEmpty()) {
+        int bestRank = Integer.MAX_VALUE;
         for (SegmentInstanceCandidate candidate : candidates) {
-          int idx = serverRankList.indexOf(candidate.getInstance());
-          if (idx == -1) {
+          // If the candidate is already selected previously, implying that it is the best candidate, select it and exit the loop.
+          if (selectedServers.contains(candidate.getInstance())) {
+            selectedInstance = candidate;
+            break;
+          }
+          Integer rank = serverRankMap.get(candidate.getInstance());
+          if (rank == null) {
```

Comment:
We want to use round-robin approach even if stats for a single server is missing. 

So we should continue to assign `selectedInstance = candidates.get(instanceIdx);`

(nit): rename `instanceIdx` to `roundRobinInstanceIdx` ?

Reply:
You are right, let me fix that